w3af/plugins/audit/open_api_auth.py

```diff
@@ -0,0 +1,372 @@
+        then the check should expect 401
+        If the API spec says a particular method is not supported,
+        then the check should expect 405 (preferred) or 401
+        (should it only expect 405 in this case?).
+      * Allow a user to specify response codes that the plugin should expect.
+        Currently the plugin expects only 401.
+        We can introduce 'expected_code_regex' configuration parameter
+        which set a regex for expected response codes.
+      * If the plugin is updated to send other HTTP methods (see above),
+        we may want to introduce another configuration parameter
+        to set expected response codes in this case.
+      * Severity of identified issues may depend on response codes
+        returned by the server.
+
+    :author: Artem Smotrakov (artem.smotrakov@gmail.com)
+    :author: Andres Riancho (andres.riancho@gmail.com)
+    """
+
+    def __init__(self):
+        AuditPlugin.__init__(self)
+        self._spec = None
+        self._expected_codes = [401]
+
+    def audit(self, freq, orig_response, debugging_id):
+        """
+        Check if an API endpoint requires authentication according to its API spec.
+
+        :param freq: A FuzzableRequest to the API endpoint.
+        :param orig_response: The HTTP response associated with the fuzzable request.
+        :param debugging_id: A unique identifier for this call to audit().
+        """
+        # The check works only with API endpoints.
+        # crawl.open_api plugin has to be called before.
+        if not self._is_api_spec_available():
+            om.out.information("Could not find an API specification, skip")
+            return
+
+        # The API spec must define authentication mechanisms.
+        if not self._has_security_definitions_in_spec():
+            om.out.information("The API spec has no security definitions, skip")
+            return
+
+        # The check only runs if the API specification
+        # requires authentication for a endpoint.
+        if not self._has_auth(freq):
+            om.out.information("Request doesn't have auth info, skip")
+            return
+
+        # Remove auth info from the request.
+        freq_with_no_auth = self._remove_auth(freq)
+
+        # Send the request to the server, and check if access denied.
+        self._uri_opener.send_mutant(freq_with_no_auth,
+                                     callback=self._check_response_code,
+                                     debugging_id=debugging_id)
+
+    def _check_response_code(self, freq, response):
+        """
+        Check if an HTTP request contains an expected response code.
+
+        :param freq: A FuzzableRequest to the API endpoint.
+        :param response: The HTTP response associated with the fuzzable request.
+        """
+        if response.get_code() not in self._expected_codes:
+            desc = 'A %s request without auth info was send to %s. ' \
```

Comment:
`... request without authentication information ...`

`was sent to`

`replied with unexpected HTTP response code %d (expected one of 401, 403, etc.)`

And add: `This is a strong indicator of a REST API authentication bypass`.